File: TeamCode/src/main/java/org/firstinspires/ftc/common/hardware/subsystems/Subsystem.java
package org.firstinspires.ftc.common.hardware.subsystems;

import com.qualcomm.robotcore.hardware.HardwareMap;

public abstract class Subsystem {
	private int readIntervalMs = 0;
	private int computeIntervalMs = 0;
	private int writeIntervalMs = 0;
	private int lastReadTimestamp = 0;
	private int lastComputeTimestamp = 0;
	private int lastWriteTimestamp = 0;

	public Subsystem() {
		SubsystemManager.getInstance().add(this);
	}

	public abstract void init(HardwareMap hwMap);
	public abstract void read();
	public abstract void compute();
	public abstract void write();

	// These are accessible to users
	public void setReadIntervalMs(int readIntervalMs) {
		this.readIntervalMs = readIntervalMs;
	}

	public void setComputeIntervalMs(int computeIntervalMs) {
		this.computeIntervalMs = computeIntervalMs;
	}

	public void setWriteIntervalMs(int writeIntervalMs) {
		this.writeIntervalMs = writeIntervalMs;
	}

	// These are only accessible to the manager
	int getReadIntervalMs() {
		return readIntervalMs;
	}

	int getComputeIntervalMs() {
		return computeIntervalMs;
	}

	int getWriteIntervalMs() {
		return writeIntervalMs;
	}

	int getLastReadTimestamp() {
		return lastReadTimestamp;
	}

	int getLastComputeTimestamp() {
		return lastComputeTimestamp;
	}

	int getLastWriteTimestamp() {
		return lastWriteTimestamp;
	}

	void setLastComputeTimestamp(int lastComputeTimestamp) {
		this.lastComputeTimestamp = lastComputeTimestamp;
	}

	void setLastReadTimestamp(int lastReadTimestamp) {
		this.lastReadTimestamp = lastReadTimestamp;
	}

	void setLastWriteTimestamp(int lastWriteTimestamp) {
		this.lastWriteTimestamp = lastWriteTimestamp;
	}
}

File: TeamCode/src/main/java/brickbot/quickstart/subsystems/Subsystem.kt
package brickbot.quickstart.subsystems

import com.qualcomm.robotcore.hardware.HardwareMap

abstract class Subsystem {
    private var readIntervalMs = 0
    private var computeIntervalMs = 0
    private var writeIntervalMs = 0
    private var lastReadTimestamp = 0
    private var lastComputeTimestamp = 0
    private var lastWriteTimestamp = 0

    constructor() {
        SubsystemManager.add(this)
    }

    abstract fun init(hwMap: HardwareMap)
    abstract fun read()
    abstract fun compute()
    abstract fun write()

    fun setReadIntervalMs(readIntervalMs: Int) {
        this.readIntervalMs = readIntervalMs
    }

    fun setComputeIntervalMs(computeIntervalMs: Int) {
        this.computeIntervalMs = computeIntervalMs
    }

    fun setWriteIntervalMs(writeIntervalMs: Int) {
        this.writeIntervalMs = writeIntervalMs
    }

    /**
     * This method is used by the SubsystemManager, do not use it.
     */
    internal fun update(timestamp: Int) {
        if (timestamp - lastReadTimestamp >= readIntervalMs) {
            read()
            lastReadTimestamp = timestamp
        }
        if (timestamp - lastComputeTimestamp >= computeIntervalMs) {
            compute()
            lastComputeTimestamp = timestamp
        }
        if (timestamp - lastWriteTimestamp >= writeIntervalMs) {
            write()
            lastWriteTimestamp = timestamp
        }
    }
}